commons/all/src/main/java/org/infinispan/commons/util/ProgressTracker.java

```diff
@@ -0,0 +1,138 @@
+package org.infinispan.commons.util;
+
+import java.time.Instant;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.commons.time.TimeService;
+
+import net.jcip.annotations.GuardedBy;
+import net.jcip.annotations.ThreadSafe;
+
+@ThreadSafe
+public final class ProgressTracker {
+
+   private static final Log log = LogFactory.getLog(ProgressTracker.class);
+
+   private final String name;
+   private final ScheduledExecutorService executor;
+   private final TimeService timeService;
+   private final long delay;
+   private final TimeUnit unit;
+
+   @GuardedBy("state")
+   private final State state;
+
+   @GuardedBy("state")
+   private ScheduledFuture<?> progression;
+
+   public ProgressTracker(String name, ScheduledExecutorService executor, TimeService timeService, long delay, TimeUnit unit) {
+      this.name = name;
+      this.executor = executor;
+      this.timeService = timeService;
+      this.state = new State();
+      this.delay = delay;
+      this.unit = unit;
+   }
+
+   public void addTasks(long value) {
+      synchronized (state) {
+         if (progression == null) {
+            progression = executor.scheduleAtFixedRate(state, delay, delay, unit);
+         }
+
+         state.addTasks(value);
+      }
+   }
+
+   public void removeTasks(long value) {
+      synchronized (state) {
+         if (progression == null)
+            throw new IllegalStateException("Task tracking not initialized");
+
+         state.addTasks(-value);
+      }
+   }
+
+   public void finishedAllTasks() {
+      synchronized (state) {
+         if (progression == null) return;
+
+         progression.cancel(true);
+         progression = null;
+         state.completed();
```

Comment:
Should we log this when progression is not set as well?

Reply:
Hrmm, I think we can remove it. I initially thought it would be useful to have the time between start and end.

Reply:
I think it has value, I wasn't suggesting to remove but rather cover the case when it is completed without adding tasks is all.

Reply:
Updated it so it will always execute the `completed()` method, and will log if it has a start time defined to compare against the current instant.